pkg/fqdn/cache.go

```diff
@@ -0,0 +1,191 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fqdn
+
+import (
+	"bytes"
+	"net"
+	"sort"
+	"time"
+
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+var DefaultDNSCache = NewDNSCache()
+
+// cacheEntry objects are immutable once created.
+// They hold an the DNS data passed in with DNSCache.Update, nominally equating
+// to a DNS lookup. cacheEntry objects are managed by DNSCache and are not
+// expected to be returned.
+type cacheEntry struct {
+	Name           string
+	LookupTime     time.Time
+	ExpirationTime time.Time
+	TTL            int
+	IPs            []net.IP
+}
+
+func (entry *cacheEntry) isExpiredBy(pointInTime time.Time) bool {
+	return pointInTime.After(entry.ExpirationTime)
+}
+
+// ipEntries maps a unique IP to the cacheEntry that provides it in .IPs.
+// Multiple IPs may point to the same cacheEntry, or they may all be different.
+// Crucially, an IP may be present in a cacheEntry but the IP in ipEntries
+// points to another cacheEntry. This is because the second cacheEntry has a
+// later expiration for this specific IP, and may not include the other IPs
+// provided by the first entry.
+// The DNS name in the entries is not checked, but is assumed to be the same
+// for all entries.
+// Note: They are guarded by the DNSCache mutex.
+type ipEntries map[string]*cacheEntry
+
+// getIPs returns a sorted list of non-expired unique IPs.
+// This needs a read-lock
+func (s ipEntries) getIPs(now time.Time) []net.IP {
+	ips := make([]net.IP, 0, len(s)) // worst case size
+	for ip, entry := range s {
+		if entry != nil && !entry.isExpiredBy(now) {
+			ips = append(ips, net.ParseIP(ip))
+		}
+	}
+
+	return keepUniqueIPs(ips) // sorts IPs
+}
+
+// DNSCache manages DNS data that will expire after a certain TTL. Information
+// is tracked per-IP address, retaining the latest-expiring DNS data for each
+// address.
+// For most real-world DNS data, the entry per name remains small because newer
+// lookups replace older ones. Large TTLs may cause entries to grow if many
+// unique IPs are returned in separate lookups.
+// Redundant or expired entries are removed on insert.
+// Lookups check for expired entries.
+type DNSCache struct {
+	lock.RWMutex
+
+	// forward DNS lookups name -> IPEntries
+	// IPEntries maps IP -> entry that provides it. An entry may provide multiple IPs.
+	forward map[string]ipEntries
+}
+
+// NewDNSCache returns an initialized DNSCache
+func NewDNSCache() *DNSCache {
+	c := &DNSCache{
+		forward: make(map[string]ipEntries),
+	}
+
+	return c
+}
+
+// Update inserts a new entry into the cache.
+// After insertion cache entries for name are expired and redundant entries
+// evicted. This is O(number of new IPs) for eviction, and O(number of IPs for
+// name) for expiration.
```

Comment:
Update documentation here that the TTL is specified in seconds

Reply:
The other parameters also needed explanation, so I added those too. Thanks